File: кучи.java
1) Бинарная куча (binary heap)

import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

public class BinaryHeap<T extends Comparable<T>> {
    private List<T> a = new ArrayList<>();

2) Биномиальная куча (binomial heap)

class BNode<T extends Comparable<T>> {
    T key;
    int degree = 0;
    BNode<T> parent, child, sibling;
    BNode(T k) { key = k; }
}

public class BinomialHeap<T extends Comparable<T>> {
    private BNode<T> head;

3) Куча Фибоначчи (Fibonacci heap)

class FibNode<T extends Comparable<T>> {
    T key;
    int degree;
    boolean mark;
    FibNode<T> parent, child, left, right;
    FibNode(T k) { key = k; left = right = this; }
}

public class FibonacciHeap<T extends Comparable<T>> {
    private FibNode<T> min;
    private int n;
template<typename T>
class FibonacciHeap {
    FNode<T>* min = nullptr;
    int n = 0;

File: кучи.cpp
1) Бинарная куча (binary heap)

#include <vector>
#include <stdexcept>

template<typename T>
class BinaryHeap {
    std::vector<T> a;
    void sift_up(int i) {
        while (i > 0) {
            int p = (i - 1) / 2;
            if (a[i] < a[p]) { std::swap(a[i], a[p]); i = p; }
            else break;
        }
    }

2) Биномиальная куча (binomial heap)

#include <iostream>
#include <memory>

template<typename T>
struct BNode {
    T key;
    int degree = 0;
    BNode* parent = nullptr;
    BNode* child = nullptr;
    BNode* sibling = nullptr;
    BNode(T k): key(k){}
};

template<typename T>
class BinomialHeap {
    BNode<T>* head = nullptr;

3) Куча Фибоначчи (Fibonacci heap)

template<typename T>
struct FNode {
    T key;
    int degree = 0;
    bool mark = false;
    FNode* parent = nullptr;
    FNode* child = nullptr;
    FNode* left = this;
    FNode* right = this;
    FNode(T k): key(k) {}
};

template<typename T>
class FibonacciHeap {
    FNode<T>* min = nullptr;
    int n = 0;